sklearn/preprocessing/_target_encoder.py

```diff
@@ -0,0 +1,337 @@
+import numpy as np
+
+from numbers import Real, Integral
+
+from ._encoders import _BaseEncoder
+from ..base import OneToOneFeatureMixin
+from ._target_encoder_fast import _fit_encoding_fast
+from ._target_encoder_fast import _fit_encoding_fast_auto_smooth
+from ..utils.validation import _check_y, check_consistent_length
+from ..utils.multiclass import type_of_target
+from ..utils._param_validation import Interval, StrOptions
+
+
+class TargetEncoder(OneToOneFeatureMixin, _BaseEncoder):
+    """Target Encoder for regression and classification targets.
+
+    Each category is encoded based on a shrinked estimate of the average target
+    values for observations belonging to the category. The encoding scheme mixes
+    the global target mean with the target mean conditioned on the value of the
+    category. [MIC]_
+
+    :class:`TargetEncoder` considers missing values, such as `np.nan` or `None`,
+    as another category and encodes them like any other category.
```

Comment:
> as of now, TargetEncoder has no explicit support for missing values and in particular encoded with `np.nan` in the original data are target encoded as any other categories.

In dirty-cat we have two parameters that handle missing values seen during fit and unknown categories seen during transform.

For missing values during fit, we added a parameter [`handle_missing`](https://github.com/dirty-cat/dirty_cat/blob/ea009b9/dirty_cat/_target_encoder.py#L49) that either imputes all the missing values as blank strings ' ' or raises an error. 

For unseen categories in transform, we have a [`handle_unknown`](https://github.com/dirty-cat/dirty_cat/blob/ea009b9/dirty_cat/_target_encoder.py#L43) parameter that either raises an error or encodes them with the prior mean of the target variable. This is mentionned in Micci-Barreca (2001) (page 30, first paragraph):
> "If missing values have no particular relationship with the dependent then S_{o} is going to converge towards a prior probability of the target which corresponds to a "neutral" representation of the missing value"
This was inspired by the `handle_missing` parameter of the [OneHotEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html) so I think users would easily understand this if you add it.

From what I see, you implement both under the hood, I guess that's what you mean @ogrisel by explicit support.
If needed to add the docstring, I think the one from dirty-cat could be adapted easily.

Reply:
A `handle_missing` parameter makes sense to me. Given the current size of this PR, I prefer to consider the `handle_missing` parameter in a follow up PR.